diff -urN node-v0.4.1.orig/tools/wafadmin/Tools/node_addon.py node-v0.4.1/tools/wafadmin/Tools/node_addon.py
--- node-v0.4.1.orig/tools/wafadmin/Tools/node_addon.py	2011-02-19 18:45:34.000000000 -0800
+++ node-v0.4.1/tools/wafadmin/Tools/node_addon.py	2011-02-19 23:55:31.928066002 -0800
@@ -28,7 +28,7 @@
   lib = join(prefix, 'lib')
 
   conf.env['LIBPATH_NODE'] = lib
-  conf.env['CPPPATH_NODE'] = join(prefix, 'include', 'node')
+  conf.env['CPPPATH_NODE'] = join(prefix, 'include', 'nodejs')
 
   conf.env.append_value('CPPFLAGS_NODE', '-D_GNU_SOURCE')
   conf.env.append_value('CPPFLAGS_NODE', '-DEV_MULTIPLICITY=0')
@@ -64,7 +64,7 @@
     nodePath = None
     if not os.environ.has_key('NODE_PATH'):
         if not os.environ.has_key('HOME'):
-            nodePath = join(get_prefix(), 'lib', 'node')
+            nodePath = join(get_prefix(), 'lib', 'nodejs')
         else:
             nodePath = join(os.environ['HOME'], '.node_libraries')
     else:
@@ -74,7 +74,7 @@
 def get_prefix():
     prefix = None
     if not os.environ.has_key('PREFIX_NODE'):
-        prefix = os.path.abspath(os.path.join(os.path.dirname(__file__), '..', '..', '..', '..'))
+        prefix = '/usr'
     else:
         prefix = os.environ['PREFIX_NODE']
     return prefix
